fix(versioning/hashicorp): better parsing

mcfedr opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

Correctly handle conversion of Hashicorp version ranges to NPM, previously multiple ranges would have been incorrectly converted, and the conversion wasnt always used.

## Context

This prevents `bump` and `replace` from correctly working for terraform modules with versions using `~>`.

currently an error like

```
DEBUG: Unsupported range type for rangeStrategy=bump: ~> 3.5 (repository=my-project)
```

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [X] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [X] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

viceice reviewed (dismissed):
LGTM

rarkins commented on lib/modules/versioning/hashicorp/convertor.spec.ts:
```diff
@@ -0,0 +1,77 @@
+  `(
+    'npm2hashicorp("$constraint") === $expected',
+    ({ expected, constraint }) => {
+      expect(npm2hashicorp(constraint)).toBe(expected);
```
Is it intentional/deliberate to split the tests this way? Ie instead of having one table and verifying both directions in one go

mcfedr replied:
Didnt think of it, but actually yes, the cases are slightly different. but I like the idea of merging the ones that are the same, and i'll keep a separate block for the extra cases.

the extras are because there are cases that can be differently represented in the different languages

rarkins commented on lib/modules/versioning/hashicorp/convertor.spec.ts:
```diff
@@ -0,0 +1,77 @@
+  `(
+    'npm2hashicorp("$constraint") === $expected',
+    ({ expected, constraint }) => {
+      expect(npm2hashicorp(constraint)).toBe(expected);
```
Is it intentional/deliberate to split the tests this way? Ie instead of having one table and verifying both directions in one go

mcfedr commented:
So I made one more change, now that the versions are properly converted, its not needed to have the specific implementation for hasihcorp of some of the `newValues`, so I've removed it and replaced with the upstream npm version. All the tests run exactly the same.

viceice commented on lib/modules/versioning/hashicorp/index.ts:
```diff
@@ -16,39 +16,36 @@ export const supportedRangeStrategies: RangeStrategy[] = [
 
-export const isValid = (input: string): boolean =>
-  !!input && npm.isValid(hashicorp2npm(input));
+export const isValid = (input: string): boolean => {
```
convert to normal function like `isLessThanRange`

viceice reviewed (commented):
otherwise LGTM

rarkins commented on lib/modules/versioning/hashicorp/convertor.spec.ts:
```diff
@@ -0,0 +1,68 @@
+  // These are non-reflective cases for npm2hashicorp
+  test.each`
+    hashicorp     | npm
+    ${'~> 4.0'}   | ${'^4'}
```
What would the new value be once 5.0.0 is released?

mcfedr replied:
These ranges both limit to 4.x.x releases

mcfedr replied:
You can see [here](https://github.com/renovatebot/renovate/pull/19652/files#diff-9f1872925310cf22ca0512e20a66c0ff19c7409c8ab0625c99dd23f1f5785ba0R68), that the `getNewValue` function will return an updated range for `~> 5.0`

mcfedr commented on lib/modules/versioning/hashicorp/index.ts:
```diff
@@ -16,39 +16,36 @@ export const supportedRangeStrategies: RangeStrategy[] = [
+  return false;
+}
 
 const matches = (version: string, range: string): boolean =>
```
No problem ))
